Lms/hash filters for pusher comparison

anathomical opened the pull request:
## WHAT
Use a hash function for the Pusher payload sent to the front-end for comparison to figure out which Admin Snapshot to reload from cache
## WHY
In cases with a lot of filters selected, sending the full payload of all filters was causing Pusher to throw errors because the payloads were too big.
## HOW
Implement identical logic for hashing the values that were being compared between the front and back ends so that the backend hashes the payload to keep Pusher payloads very small and then the front-end can confirm if that hash matches the payload it expects as a signal to reload data from cache.

### Notion Card Links
https://www.notion.so/quill/Hash-Admin-Snapshot-filter-payloads-on-the-front-end-and-back-end-for-Pusher-comparison-instead-of-t-6b1ae512decd4d13bfd5425e24a33e4a?pvs=4
https://www.notion.so/quill/Pusher-Payload-Too-Large-Errors-4f82617b954f46999ca623acd6cb0984?pvs=4

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Yes
Self-Review: Have you done an initial self-review of the code below on Github? | Yes
Spec Review: Have you reviewed the spec and ensured this meets requirements and/or matches design mockups? | (N/A or Yes)

anathomical committed:
Test hash comparison for Pusher payloads

anathomical committed:
Don't change package.json for md5

anathomical committed:
Install md5 npm package in root

anathomical committed:
Try npm install with npm@8.5.5

anathomical committed:
Use npm install --lockfile-version 1 to see if that works

anathomical committed:
Use safe accessors when joining data for md5 hashing

anathomical committed:
Tweak joins to hopefully hash identical strings

anathomical committed:
Add payload info

anathomical committed:
Add more logging

anathomical committed:
Finalize hashing of keys passed through Pusher

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into lms/hash-filters-for-pusher-comparison

anathomical committed:
Update specs to cover expectation of hash messages in Pusher

anathomical committed:
Make a small tweak to simplify spec setup

dandrabik commented on services/QuillLMS/app/workers/snapshots/cache_snapshot_count_worker.rb:
```diff
@@ -35,13 +35,16 @@ def perform(cache_key, query, user_id, timeframe, school_ids, filters)
+        filters['grades']&.join('-'),
+        filters['teacher_ids']&.join('-'),
+        filters['classroom_ids']&.join('-')
+      ].join('-'))
```
Small refactor, make an array, flatten in, then you can do the join once
```ruby
params = [query, timeframe['name'], school_ids, filters['grades'], filters['teacher_ids'], filters['classroom_ids']]

filter_hash = params
  .flatten
  .join('-')

dandrabik commented on services/QuillLMS/app/workers/snapshots/cache_snapshot_top_x_worker.rb:
```diff
@@ -23,13 +23,16 @@ def perform(cache_key, query, user_id, timeframe, school_ids, filters)
-        }.merge(filters)
-      )
+      filter_hash = Digest::MD5.hexdigest([
+        query,
```
Same comment, here, we can do this `join('-')` once.

dandrabik commented on services/QuillLMS/client/app/bundles/PremiumHub/components/usage_snapshots/snapshotCount.tsx:
```diff
@@ -1,5 +1,7 @@
 import * as React from 'react'
 
+import md5 from 'md5'
```
I believe that JS has a [native digest function](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest). Let's use that if possible.

anathomical replied:
Per our discussion on Slack, the difficulty of getting native stuff to work both in-browser and in test environment, I'm sticking with a separate md5 library for now.

dandrabik commented on services/QuillLMS/client/app/bundles/PremiumHub/components/usage_snapshots/snapshotCount.tsx:
```diff
@@ -111,16 +113,18 @@ const SnapshotCount = ({ label, size, queryKey, searchCount, selectedGrades, sel
-      if (queryKeysAreEqual && timeframesAreEqual && schoolIdsAreEqual && gradesAreEqual && teacherIdsAreEqual && classroomIdsAreEqual) {
-        getData()
-      }
+      const filterHash = md5(filterTarget)
```
Can we extract this out into a function to make it testable. I'd like you to add a test for this.

dandrabik commented on services/QuillLMS/client/app/bundles/PremiumHub/components/usage_snapshots/snapshotCount.tsx:
```diff
@@ -111,16 +113,18 @@ const SnapshotCount = ({ label, size, queryKey, searchCount, selectedGrades, sel
+      const filterTarget = [
+        queryKey,
+        selectedTimeframe,
+        selectedSchoolIds.join('-'),
```
Would have to do equivalent change on the frontend.

dandrabik commented on services/QuillLMS/app/workers/snapshots/cache_snapshot_count_worker.rb:
```diff
@@ -35,13 +35,16 @@ def perform(cache_key, query, user_id, timeframe, school_ids, filters)
-          school_ids: school_ids
-        }.merge(filters)
-      )
+      filter_hash = Digest::MD5.hexdigest([
```
Let's make this a separate service and test it explicitly with generic data, e.g.
```ruby
digest = HashDigest.run(array)

dandrabik reviewed (changes requested):
Thanks for your work on this, I think this is getting close. There are a few small changes in my comments, but the larger item is adding better tests (and making the code testable):

1. Use the native JS digest library if possible.
2. Break out these digest calculations into services / functions and write explicit unit tests for them (both in Ruby and js ), e.g. confirm that HashService.run(['hello','world]) == "adsfasdasdfadsf"
3. Include the same sample tests in Ruby and Javascript, e.g. pass this array, get this explicit hash back. We need these outputs to match, and there's not a great way to link these tests between js and ruby, so let's just copy/paste the same example input/outputs in the tests.

anathomical committed:
Move payload hashing into a separate service

anathomical committed:
Move hashing logic into its own function

And add unit test to let us confirm that matching inputs on both the front and back end produce the same outputs

anathomical committed:
Go back to using MD5 for our hash algorithm

anathomical committed:
Move payload hasing into a shared function

And add unit tests to ensure that it outputs the same hash as we get on the backend

anathomical committed:
Add some comments about identical output expectations

anathomical committed:
Move hash checking for Pusher payloads into a separate function

For better testing

anathomical committed:
Add matching "real world" tests to both front-end and back-end

dandrabik commented on services/QuillLMS/app/services/payload_hasher.rb:
```diff
@@ -0,0 +1,14 @@
+class PayloadHasher < ApplicationService
+  attr_reader :payload
+
+  def initialize(payload)
```
One small naming nit, `payload` isn't a generic param, since you are doing `.join('-')` it needs to be an `array`, you might want to name it as such.

dandrabik commented on services/QuillLMS/client/app/bundles/PremiumHub/shared.tsx:
```diff
@@ -44,3 +46,9 @@ export function mapItemsIfNotAll(selectedItems, allItems, mapKey = 'id') {
+
+/* Should produce the same output as backend `app/services/payload_hasher.rb:run` */
+export function hashPayload(payload) {
+  const joinedPayload = payload.join('-')
```
same here, you might want a more specific name for `payload` or type it, since this is typescipt.

dandrabik reviewed (approved):
Small naming nit, but overall, looks much better. Thanks for doing this refactor and adding the tests. This feels much more solid, well done.

anathomical committed:
Rename function params to make the array requirement clearer

anathomical committed:
Add some type-checking to the new front-end utility function

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into lms/hash-filters-for-pusher-comparison
